src/Translator.php

```diff
@@ -0,0 +1,542 @@
+
+    private array $modulePaths = [];
+
+    private array $originalJson = [];
+
+    private array $originalYaml = [];
+
+    private array $pullRequestUrls = [];
+
+    public function run()
+    {
+        $this->checkEnv();
+        $this->setModulePaths();
+        if ($this->doTransifexPullAndUpdate) {
+            $this->log('Updating translations for ' . count($this->modulePaths) . ' module(s)');
+            $this->storeJson();
+            $this->storeYaml();
+            $this->setJsonAndYmlFileTimes();
+            $this->transifexPullSource();
+            $this->mergeYaml();
+            $this->cleanYaml();
+            $this->mergeJson();
+            $this->collectStrings();
+            $this->generateJavascript();
+        }
+        if ($this->doTransifexPush) {
+            $this->transifexPushSource();
+        }
+        $this->gitCommitPushAndPullRequest();
+        $this->outputPullRequestUrls();
+    }
+
+    private function checkEnv(): void
+    {
+        $txInstructions = 'Install the new go version of the client https://developers.transifex.com/docs/cli';
+        if ($this->exec('which tx') === '') {
+            throw new Exception("Could not find tx executable. $txInstructions");
+        }
+        $help = $this->exec('tx help');
+        $help = str_replace(["\n", ' '], '', $help);
+        preg_match('#VERSION:([0-9]+\.[0-9]+)#', $help, $matches);
+        if (($matches[1] ?? 0) < 1.6) {
+            throw new Exception("Your version of tx is too old. $txInstructions");
+        }
+        if ($this->exec('which wget') === '') {
+            throw new Exception('Could not find wget command. Please install it.');
+        }
+        $this->txSite = getenv('TX_SITE');
+        if (!$this->txSite) {
+            throw new Exception('TX_SITE environment variable is not defined');
+        }
+        if (strpos($this->txSite, 'http://') !== 0 && strpos($this->txSite, 'https://') !== 0) {
+            $this->txSite = 'http://' . $this->txSite;
+        }
+        if (!filter_var($this->txSite, FILTER_VALIDATE_URL)) {
+            throw new Exception('SITE environment variable is not a valid url');
+        }
+        $this->githubToken = getenv('GITHUB_API_TOKEN');
+        if (!$this->githubToken) {
+            $message = implode(' ', [
+                'Could not get a valid token from GITHUB_API_TOKEN environment variable.',
+                'Create a new token with the public_repo checkbox ticked that will allow you to create pull-requests.'
+            ]);
+            throw new Exception($message);
+        }
+        // This will be set to true by default if TX_PULL environment variable is not defined
+        $this->doTransifexPullAndUpdate = !in_array(strtolower(getenv('TX_PULL')), ['false', '0']);
+        // This will be set to false by default if TX_PUSH environment variable is not defined
+        $this->doTransifexPush = in_array(strtolower(getenv('TX_PUSH')), ['true', '1']);
```

Comment:
I don't know, we have poor instructions in confluence and cow.  In a normal operation you probably want to push as well.  I suspect strongly that the i18nTextCollectorTask is supposed to be run as part of a 'push' operation and it weird in the 'pull block' But I don't fully understand what all things that I copied from cow do.

Reply:
I think it's bad configuration to have both as false.
In that scenario, you'll end up just running:
```php
$this->checkEnv();
$this->setModulePaths();
$this->gitCommitPushAndPullRequest();
$this->outputPullRequestUrls();
```
which, because there's no changes, won't do anything... someone who accidentally set the wrong config or forgot to change one of the env variables might then think they did it right, but there just weren't any changes to make.

We should throw an exception if both of these are false, to avoid that scenario.